polyfills/fm/client/fmradio.js

```diff
@@ -0,0 +1,421 @@
+//               }
+//             }
+// * Answer: { id: requestId,
+//             result|error: Whatever }
+
+// For all the invoke operations:
+//  * Request: { id: requestId,
+//               data: {
+//                 operation: methodName,
+//                 params: parameters (optional)
+//               }
+//             }
+// * Answer: Let's assume that the request is handled properly
+
+// For the onpropertychange operation
+//  Request: { id: requestId,
+//             data: {
+//             operation: 'onpropertychange',
+//             handler: desiredPropertyChange,
+//             property: propertyToUpdatedWhenTriggered
+//           },
+// Answer:
+//    { id: requestId,
+//      data: {
+//        result: {
+//          handler: desiredPropertyChange,
+//          propertyValue: newValueOfTheProperty
+//        }
+//      }
+//    }
+
+
+(function(window) {
+
+  function debug(t) {
+    console.log(t);
+  }
+
+  if (window.navigator.mozFMRadio) {
+    // Hmm it's already available... so let's just use it and be done with it
+    return;
+  }
+
+  // Wishful thinking at the moment...
+  const FMRADIO_SERVICE = 'https://fmradioservice.gaiamobile.org';
+
+  // It's nice being monothread...
+  var _currentRequestId = 1;
+
+  var serializeDOMRequest = function(data) {
+    var self = this;
+    return {
+      id: ++_currentRequestId,
+      data: data,
+      processAnswer: function(answer) {
+        if (answer.error) {
+          self._fireError(answer.error);
+        } else {
+          self._fireSuccess(answer.result);
+        }
+      }
+    };
+  };
+
+  var serializeRequest = function(data) {
+    var self = this;
+    return {
+      id: ++_currentRequestId,
+      data: data,
+      processAnswer: function(answer) {
+        if (answer.error) {
+          console.error(answer.error);
+        } else {
+          self[_getPropertyName(answer.result.name)] = answer.result.value;
+ 
+          // This is a trick beacuse the current frequency could be null if the
+          // radio is not enabled and we need to draw the UI with an initial
+          // frequency
+          if (answer.result.name === 'frequency' &&
+            answer.result.value === null) {
+            self[_getPropertyName(answer.result.name)] =
+              self.frequencyLowerBound;
+          }
+
+          var onchange = '_on' + answer.result.name.toLowerCase() + 'change';
+          // Trigger callback in order to force the update of the property value
+          if (typeof self[onchange] === 'function') {
+            self[onchange]({
+              name: 'on' + answer.result.name.toLowerCase() + 'change',
+              value: answer.result.value
+            });
+          }
+        }
+      }
+    };
+  };
+
+  var _getPropertyName = function(name) {
+    return 'real' + name.charAt(0).toUpperCase() + name.slice(1);
+  }
+
+  var _createAndQueueDOMRequest = function(data) {
+    var request = new FakeDOMRequest();
```

Comment:
The FakeDOMRequest constructor has changed! (and the next line is not needed anymore)